safekeeper/src/broker.rs

```diff
@@ -44,6 +47,117 @@ fn timeline_safekeeper_path(
     )
 }
 
+pub struct Election {
+    pub election_name: String,
+    pub candidate_name: String,
+    pub broker_endpoints: Vec<Url>,
+}
+
+impl Election {
+    pub fn new(election_name: String, candidate_name: String, broker_endpoints: Vec<Url>) -> Self {
+        Self {
+            election_name,
+            candidate_name,
+            broker_endpoints,
+        }
+    }
+}
+
+pub struct ElectionLeader {
+    client: Client,
+    keep_alive: JoinHandle<Result<()>>,
+}
+
+impl ElectionLeader {
+    pub async fn check_am_i(
+        &mut self,
+        election_name: String,
+        candidate_name: String,
+    ) -> Result<bool> {
+        let resp = self.client.leader(election_name).await?;
+
+        let kv = resp.kv().ok_or(anyhow!("failed to get leader response"))?;
+        let leader = kv.value_str()?;
+
+        Ok(leader == candidate_name)
+    }
+
+    pub async fn give_up(&self) {
+        self.keep_alive.abort();
+        // TODO: it'll be wise to resign here but it'll happen after lease expiration anyway
+        // should we await for keep alive termination?
+        // self.keep_alive.await;
+    }
+}
+
+pub async fn get_leader(req: &Election) -> Result<ElectionLeader> {
+    let mut client = Client::connect(req.broker_endpoints.clone(), None)
+        .await
+        .context("Could not connect to etcd")?;
+
+    let lease = client
+        .lease_grant(LEASE_TTL_SEC, None)
+        .await
+        .context("Could not acquire a lease");
+
+    let lease_id = lease.map(|l| l.id()).unwrap();
+
+    let keep_alive = spawn::<_>(lease_keep_alive(client.clone(), lease_id));
+
+    if let Err(e) = client
+        .campaign(
+            req.election_name.clone(),
+            req.candidate_name.clone(),
+            lease_id,
+        )
+        .await
+    {
+        keep_alive.abort();
+        return Err(e.into());
+    }
+
+    Ok(ElectionLeader { client, keep_alive })
+}
+
+pub async fn lease_keep_alive(mut client: Client, lease_id: i64) -> Result<()> {
+    let (mut keeper, mut ka_stream) = client
+        .lease_keep_alive(lease_id)
+        .await
+        .context("failed to create keepalive stream")?;
+
+    loop {
+        let push_interval = Duration::from_millis(PUSH_INTERVAL_MSEC);
+
+        keeper
+            .keep_alive()
+            .await
+            .context("failed to send LeaseKeepAliveRequest")?;
```

Comment:
Is there a way we are now able to react on these errors?
Presumably, we are not the leader anymore after this happens, but where is the logic to cover the leadership loss?

We never log these events and never really join on this task, that seems quite suspicious.
I wonder, if `check_am_i` and keepalive has to be the same thing, running in `select!` somewhere?

----

Another reason for being suspicious is the `WalBackupTask::run` workflow, where the keepalive task is being spawned and killed constantly in a loop.

Reply:
As soon as lease is lost, next call to `check_am_i` will tell our leadership is lost. Then we jump out of offloading loop into election loop and destroy the lease task (if it hasn't terminated itself at this point). But again, I feel like we could get away without Election API at all later.

Reply:
Due to compiler quirks: by default, the most "owning" resolution is used, similar to 

```rust
struct Foo {
    field: String,
}

fn main() {
    let foo = Foo {
        field: "field".to_string(),
    };

    let foo_ref_field_1 = &foo.field;
    let foo_ref_field_2 = &foo.field;

    // moves out the value
    let foo_field_1 = foo.field;
    // error, cannot move twice
    let foo_field_2 = foo.field;
}
```

so if you rewrite `async fn join(&mut self) ` contents as 

```
async fn join(&mut self) {
        let zz = &mut self.h;
        zz.await;
    }
```

it  would work.